File: 1. TESTE DE WEB SCRAPING/scraping_python/src/downloader/pdf_downloader.py
import os
import re
import requests
from bs4 import BeautifulSoup

class PDFDownloader:
    def __init__(self, url, download_dir, anexos):
        self.url = url
        self.download_dir = download_dir
        self.anexos = anexos

    def download_anexos(self):
        BLUE = "\033[94m"
        GREEN = "\033[92m"
        RED = "\033[91m"
        RESET = "\033[0m"

        print(f"{BLUE}Acessando: {self.url}{RESET}")
        try:
            response = requests.get(self.url)
            response.raise_for_status()
            soup = BeautifulSoup(response.text, 'html.parser')

            links = soup.find_all('a', href=True)
            downloaded = []

            for anexo in self.anexos:
                nome = anexo["nome"]
                pattern = re.compile(anexo["regex"], re.IGNORECASE)

                for link in links:
                    href = link.get("href", "")
                    text = link.get_text()

                    if href.endswith(".pdf") and (pattern.search(text) or pattern.search(href)):
                        full_url = href
                        if not href.startswith("http"):
                            full_url = f"https://www.gov.br{href}" if href.startswith("/") else f"https://www.gov.br/{href}"

                        file_path = os.path.join(self.download_dir, f"{nome.replace(' ', '_')}.pdf")

                        if self.download_file(full_url, file_path):
                            downloaded.append(file_path)
                        break  
            return downloaded

        except Exception as e:
            print(f"{RED}Erro ao acessar o site: {e}{RESET}")
            return []

    def download_file(self, url, filepath):
        BLUE = "\033[94m"
        GREEN = "\033[92m"
        RED = "\033[91m"
        RESET = "\033[0m"

        try:
            print(f"{BLUE}Baixando: {url}{RESET}")
            response = requests.get(url, stream=True)
            response.raise_for_status()

            with open(filepath, 'wb') as f:
                for chunk in response.iter_content(chunk_size=8192):
                    f.write(chunk)

            print(f"{GREEN}Download concluído: {filepath}{RESET}")
            return True
        except Exception as e:
            print(f"{RED}Erro ao baixar {url}: {e}{RESET}")
            return False

File: 1. TESTE DE WEB SCRAPING/scraping_java/src/main/java/com/scraping/java/directory/PDFDownloader.java
package com.scraping.java.directory;

import com.scraping.java.LogsMain;
import com.scraping.java.config.Anexo;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class PDFDownloader {

    private final String url;
    private final String downloadDir;
    private final List<Anexo> anexos;

    public PDFDownloader(String url, String downloadDir, List<Anexo> anexos) {
        this.url = url;
        this.downloadDir = downloadDir;
        this.anexos = anexos;
    }

    public List<String> downloadAnexos() throws IOException {
        LogsMain.logWarning("Acessando: " + url);
        List<String> downloaded = new ArrayList<>();
        Document doc = Jsoup.connect(url).get();
        Elements links = doc.select("a[href$=.pdf]");

        for (Anexo anexo : anexos) {
            Pattern pattern = Pattern.compile(anexo.regex, Pattern.CASE_INSENSITIVE);
            for (Element link : links) {
                String linkText = link.text();
                String href = link.attr("href");
                if (pattern.matcher(linkText).find() || pattern.matcher(href).find()) {
                    String fullUrl = href.startsWith("http") ? href : "https://www.gov.br" + (href.startsWith("/") ? href : "/" + href);
                    String filename = downloadDir + File.separator + anexo.nome.replaceAll(" ", "_") + ".pdf";
                    if (downloadFile(fullUrl, filename)) {
                        downloaded.add(filename);
                    }
                    break;
                }
            }
        }
        return downloaded;
    }

    private boolean downloadFile(String fileURL, String savePath) {
        try (BufferedInputStream in = new BufferedInputStream(new URL(fileURL).openStream());
             FileOutputStream fileOutputStream = new FileOutputStream(savePath)) {
            byte[] dataBuffer = new byte[1024];
            int bytesRead;
            while ((bytesRead = in.read(dataBuffer, 0, 1024)) != -1) {
                fileOutputStream.write(dataBuffer, 0, bytesRead);
            }
            LogsMain.logInfo("Download concluído: " + savePath);
            return true;
        } catch (IOException e) {
            LogsMain.logInfo("Erro ao baixar: " + fileURL + " - " + e.getMessage());
            return false;
        }
    }
}
